sdk/js: Remove Solana WASM Dependencies

a5-pickle opened the pull request:
### Overview
The purpose of this PR is to remove any Solana WASM dependencies. In the existing Javascript SDK, WASMs are needed for the following:
* Importing methods to create Solana program instructions (e.g. `transfer_native_ix`)
* Importing methods to generate PDAs (e.g. `guardian_set_address`)
* Importing methods to get account data (e.g. `parse_guardian_set`)
* When running a Node.js script, you would have to call `setDefaultWasm` before you can use any of the Solana Wormhole SDK methods that require Solana program instructions
* To parse a VAA (verified wormhole message), you would use `parse_vaa` from `importCoreWasm`, which is silly because the deserialization should just exist by parsing a Buffer

Even though these seemed convenient to have (by having the WASM directly connected to the Rust code that makes up the Solana programs), having to import WASMs everywhere to get things done can be annoying for integrators. And if an integrator just cares about the instructions, PDA derivation and account deserialization, no need to `await` for anything now.

### How to Review this PR

Before you sit down and review all the changes, make sure you use the bathroom because this is a doozy of a PR. It wasn't particularly fun putting these changes in, either. But I apologize for having to share my pain with you.

These are the main parts of the PR that you need to review:
* `solana`: Added IDLs for Wormhole (Core Bridge), Token Bridge and NFT Bridge in the `idl` directory. Only the instructions were added to this IDL.
* `sdk/js`: Start with the `src/solana` directory to see how I wrote up the specific IDL coders in the `wormhole`, `tokenBridge` and `nftBridge` directories. I've also added things for convenience in the `utils` directory, which a lot of the things I've written use; review this if you're curious, but they are written mostly to perform what I needed to replace the WASM-dependent methods and create tests (interacting w/ the secp Solana program to verify multiple EC signatures is my fav).
* `sdk/js`: In each of `token_bridge` and `nft_bridge` directories of `src`, Solana-specific methods now use the WASM-less instructions, PDA derivation and account deserialization methods. If you took the diff of `token_bridge/attest.ts` and looked at `attestFromSolana` for example, you'll see that we now use `createAttestTokenInstruction` as opposed to `attest_ix` which required us to use `importTokenWasm()`.
* `testing`: Added a directory called `solana_test_validator` which is not a part of the Github workflow (would like to add it, but was not sure how to do it). I wrote all of these tests from scratch to verify everything I've changed and added to the SDK. These verify ALL instructions (including governance-based ones like `upgrade_contract`), PDA derivation, CPI accounts and account data getters. This test does not require Tilt because it spins up a Solana local validator and deploys all of the relevant programs.

If you have the Solana CLI on your machine, I would checkout this branch (`sdk/remove-wasm`) and go to `testing/solana-test-validator` and run `make test` to run through all of the SDK tests. Least of all this should run and all tests should pass.

Then go through the `sdk/js` directory in the order I specified above to get a sense of what I needed to write.

Finally you can go through the SDK tests I wrote to see if I missed anything.

Please let me know if you have any trouble reviewing this PR.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/certusone/wormhole/1375)
<!-- Reviewable:end -->

kcsongor commented on sdk/js/src/nft_bridge/__tests__/integration.ts:
```diff
@@ -59,6 +59,7 @@ import {
+import { parseVaa } from "../../vaa/wormhole";
 const ERC721 = require("@openzeppelin/contracts/build/contracts/ERC721PresetMinterPauserAutoId.json");
 
 setDefaultWasm("node");
```
can this be removed too?

karlkempe replied:
`setDefaultWasm` will eventually be removed once the Token Bridge and NFT Bridge wasm imports are completely swapped out with the wasmless calls. Soon!

kcsongor replied:
What a day it will be!

kcsongor commented:
Hey @a5-pickle, I'm starting to look over this PR. It's quite big, could you add a brief overview into the PR description about what it does and what's a good way to review it? 🙏

a5-pickle commented:
> Hey @a5-pickle, I'm starting to look over this PR. It's quite big, could you add a brief overview into the PR description about what it does and what's a good way to review it? 🙏

I should have written a comment when I initially opened a PR!

### Overview
The purpose of this PR is to remove any Solana WASM dependencies. In the existing Javascript SDK, WASMs are needed for the following:
* Importing methods to create Solana program instructions (e.g. transfer_native_ix)
* Importing methods to generate PDAs (e.g. `guardian_set_address`)
* Importing methods to get account data (e.g. `parse_guardian_set`)
* When running a Node.js script, you would have to call `setDefaultWasm` before you can use any of the Solana Wormhole SDK methods that require Solana program instructions
* To parse a VAA (verified wormhole message), you would use `parse_vaa` from `importCoreWasm`, which is silly because the deserialization should just exist by parsing a Buffer

Even though these seemed convenient to have (by having the WASM directly connected to the Rust code that makes up the Solana programs), having to import WASMs everywhere to get things done can be annoying for integrators. And if an integrator just cares about the instructions, PDA derivation and account deserialization, no need to `await` for anything now.

### How to Review this PR

Before you sit down and review all the changes, make sure you use the bathroom because this is a doozy of a PR. It wasn't particularly fun putting these changes in, either. But I apologize for having to share my pain with you.

There are three main parts of the PR that you need to review:
* `solana`: Added IDLs for Wormhole (Core Bridge), Token Bridge and NFT Bridge in the `idl` directory. Only the instructions were added to this IDL.
* `sdk/js`: Start with the `src/solana` directory to see how I wrote up the specific IDL coders in the `wormhole`, `tokenBridge` and `nftBridge` directories. I've also added things for convenience in the `utils` directory, which a lot of the things I've written use; review this if you're curious, but they are written mostly to perform what I needed to replace the WASM-dependent methods and create tests (interacting w/ the secp Solana program to verify multiple EC signatures is my fav).
* `sdk/js`: In each of `token_bridge` and `nft_bridge` directories of `src`, Solana-specific methods now use the WASM-less instructions, PDA derivation and account deserialization methods. If you took the diff of `token_bridge/attest.ts` and looked at `attestFromSolana` for example, you'll see that we now use `createAttestTokenInstruction` as opposed to `attest_ix` which required us to use `importTokenWasm()`.
* `testing`: Added a directory called `solana_test_validator` which is not a part of the Github workflow (would like to add it, but was not sure how to do it). I wrote all of these tests from scratch to verify everything I've changed and added to the SDK. These verify ALL instructions (including governance-based ones like `upgrade_contract`), PDA derivation, CPI accounts and account data getters. This test does not require Tilt because it spins up a Solana local validator and deploys all of the relevant programs.

If you have the Solana CLI on your machine, I would checkout this branch (`sdk/remove-wasm`) and go to `testing/solana-test-validator` and run `make test` to run through all of the SDK tests. Least of all this should run and all tests should pass.

Then go through the `sdk/js` directory in the order I specified above to get a sense of what I needed to write.

Finally you can go through the SDK tests I wrote to see if I missed anything.

Please let me know if you have any trouble reviewing this PR.

a5-pickle commented:
I just rebased w/ the latest dev.v2 commit and now Tilt sdk-ci-tests is getting the following error:
```bash
2022-10-25T04:44:18.7287970Z  sdk-ci-tests │   console.log
2022-10-25T04:44:18.7288343Z  sdk-ci-tests │     Creating wallet...
2022-10-25T04:44:18.7288655Z  sdk-ci-tests │
2022-10-25T04:44:18.7289054Z  sdk-ci-tests │       at src/cosmwasm/query.testnet.test.ts:322:13
2022-10-25T04:44:18.7289407Z  sdk-ci-tests │
2022-10-25T04:44:18.7679602Z  sdk-ci-tests │   console.error
2022-10-25T04:44:18.7680318Z  sdk-ci-tests │     Error: the mnemonic contains a word that is not in the wordlist
2022-10-25T04:44:18.7680997Z  sdk-ci-tests │         at seedFromMnemonic (/app/sdk/js/node_modules/algosdk/src/mnemonic/mnemonic.ts:106:42)
2022-10-25T04:44:18.7681868Z  sdk-ci-tests │         at Object.mnemonicToSecretKey (/app/sdk/js/node_modules/algosdk/src/mnemonic/mnemonic.ts:147:16)
2022-10-25T04:44:18.7682490Z  sdk-ci-tests │         at /app/sdk/js/src/cosmwasm/query.testnet.test.ts:323:33
2022-10-25T04:44:18.7683043Z  sdk-ci-tests │         at step (/app/sdk/js/src/cosmwasm/query.testnet.test.ts:33:23)
2022-10-25T04:44:18.7683659Z  sdk-ci-tests │         at Object.next (/app/sdk/js/src/cosmwasm/query.testnet.test.ts:14:53)
2022-10-25T04:44:18.7684227Z  sdk-ci-tests │         at /app/sdk/js/src/cosmwasm/query.testnet.test.ts:8:71
2022-10-25T04:44:18.7684703Z  sdk-ci-tests │         at new Promise (<anonymous>)
2022-10-25T04:44:18.7685546Z  sdk-ci-tests │         at Object.<anonymous>.__awaiter (/app/sdk/js/src/cosmwasm/query.testnet.test.ts:4:12)
2022-10-25T04:44:18.7686145Z  sdk-ci-tests │         at Object.<anonymous> (/app/sdk/js/src/cosmwasm/query.testnet.test.ts:313:42)
2022-10-25T04:44:18.7686690Z  sdk-ci-tests │         at Promise.then.completed (/app/sdk/js/node_modules/jest-circus/build/utils.js:390:28)
2022-10-25T04:44:18.7687019Z  sdk-ci-tests │
2022-10-25T04:44:18.7689294Z  sdk-ci-tests │       623 |     }
2022-10-25T04:44:18.7689709Z  sdk-ci-tests │       624 |   } catch (e) {
2022-10-25T04:44:18.7689991Z  sdk-ci-tests │     > 625 |     console.error(e);
2022-10-25T04:44:18.7690255Z  sdk-ci-tests │           |             ^
2022-10-25T04:44:18.7690478Z  sdk-ci-tests │       626 |   }
2022-10-25T04:44:18.7690715Z  sdk-ci-tests │       627 | });
2022-10-25T04:44:18.7690969Z  sdk-ci-tests │       628 |
2022-10-25T04:44:18.7691324Z  sdk-ci-tests │
2022-10-25T04:44:18.7691790Z  sdk-ci-tests │       at src/cosmwasm/query.testnet.test.ts:625:13
2022-10-25T04:44:18.7692163Z  sdk-ci-tests │       at step (src/cosmwasm/query.testnet.test.ts:33:23)
2022-10-25T04:44:18.7692671Z  sdk-ci-tests │       at Object.next (src/cosmwasm/query.testnet.test.ts:14:53)
2022-10-25T04:44:18.7693036Z  sdk-ci-tests │       at src/cosmwasm/query.testnet.test.ts:8:71
2022-10-25T04:44:18.7693427Z  sdk-ci-tests │       at Object.<anonymous>.__awaiter (src/cosmwasm/query.testnet.test.ts:4:12)
2022-10-25T04:44:18.7693829Z  sdk-ci-tests │       at Object.<anonymous> (src/cosmwasm/query.testnet.test.ts:313:42)
```

@kcsongor @evan-gray May need some help debugging this one.

kcsongor commented:
FWIW if you want to add a new CI action, you can do so by editing this file: https://github.com/wormhole-foundation/wormhole/blob/dev.v2/.github/workflows/build.yml

kcsongor commented on testing/solana-test-validator/run_sdk_tests.sh:
```diff
@@ -0,0 +1,38 @@
+    echo "solana-test-validator already running"
+    exit 1;
+fi
+
```
maybe `set -e` here? then the script will error out if any of the commands error; that's useful if you later want to rely on the result of this script in CI or something

a5-pickle replied:
Because this script has a lot of background processes (`solana-test-validator` and `solana logs` for each program), wouldn't those continue running if `ts-mocha` had a !=0 exit code?

kcsongor commented on sdk/js/src/solana/wormhole/accounts/guardianSet.ts:
```diff
@@ -0,0 +1,87 @@
+}
+
+export class GuardianSetData {
+  // pub struct GuardianSetData {
```
commented out code

a5-pickle replied:
Removed this and other spots: https://github.com/wormhole-foundation/wormhole/pull/1375/commits/aaa480b0b721220f307a96467bafcea3e6b75172

kcsongor commented on sdk/js/src/solana/wormhole/accounts/config.ts:
```diff
@@ -0,0 +1,89 @@
+}
+
+export class BridgeConfig {
+  // pub struct BridgeConfig {
```
do we want to keep this commented out code?

a5-pickle replied:
Ah, these were just referencing the structs of accounts if anyone wanted to look at the source. But I can take these out

a5-pickle commented on sdk/js/src/utils/parseVaa.ts:
```diff
@@ -15,51 +14,37 @@ export const METADATA_REPLACE = new RegExp("\u0000", "g");
-  targetChain: arr.readUInt16BE(99) as ChainId,
-  fee: BigNumber.from(arr.slice(101, 101 + 32)).toBigInt(),
-});
+export function parseTransferPayload(payload: Buffer): any {
```
I kept those as any because that is what they returned before. I really wanted to move these methods entirely, but I wasn't sure who else used these methods. So I kept these here and preserved the `any` return type.

kcsongor replied:
before the return type was unspecified which meant that it was inferred based on the definition, i.e. (copied from my editor):
```typescript
const parseTransferPayload: (arr: Buffer) => {
    amount: bigint;
    originAddress: string;
    originChain: ChainId;
    targetAddress: string;
    targetChain: ChainId;
    fee: bigint;
}
```

I think even if we don't add an explicit return type, we should just keep that behaviour

a5-pickle replied:
Fixed. Also added `fromAddress` for the `parseTransferPayload` method because the old method would have assumed the from address is the fee for payload 3.

evan-gray reviewed (approved):
My approach - passes tests and is on a minor version bump with the right signatures, so it should be patch fixable if the tests for some reason don't hit enough of the coverage.

evan-gray reviewed (dismissed):
Change since last review looks fine

evan-gray reviewed (approved):
yes, was just about to double-back my comment with this:
> do not `as` parsed chain ids! this could be a number that is not yet in the type. users should coalesce after to get the type if the need it explicitly
